modules/broker-core/src/main/java/org/wso2/broker/core/util/MessageTracer.java

```diff
@@ -0,0 +1,207 @@
+    }
+
+    public static void trace(String description, TraceField... traceFields) {
+        if (LOGGER.isTraceEnabled()) {
+            TraceBuilder traceBuilder = new TraceBuilder();
+            for (TraceField traceField : traceFields) {
+                traceBuilder.field(traceField);
+            }
+            LOGGER.trace(traceBuilder.buildTrace(description));
+        }
+    }
+
+    public static boolean isTraceEnabled() {
+        return LOGGER.isTraceEnabled();
+    }
+
+    /**
+     * Internal trace message builder class.
+     */
+    private static class TraceBuilder {
+
+        private static final String FIELD_SUFFIX = " } ";
+        private static final String FIELD_PREFIX = " { ";
+        private static final String FIELD_DELIMITER = " , ";
+
+        private static final String INTERNAL_ID = "id: ";
+        private static final String EXCHANGE_NAME = "exchangeName: ";
+        private static final String ROUTING_KEY = "routingKey: ";
+        private static final String QUEUE_NAME = "queueName: ";
+        private static final String CONSUMER_ID = "consumerId: ";
+        private static final String REDELIVERY_COUNT = "redeliveryCount: ";
+        private static final Object REDELIVERED = "isRedelivered: ";
+
+        private final ArrayList<String> fields;
+
+        TraceBuilder() {
+            this.fields = new ArrayList<>();
+        }
+
+        TraceBuilder internalId(long internalId) {
+            fields.add(INTERNAL_ID + String.valueOf(internalId));
+            return this;
+        }
+
+        TraceBuilder exchangeName(String name) {
+            fields.add(EXCHANGE_NAME + String.valueOf(name));
+            return this;
+        }
+
+        TraceBuilder routingKey(String routingKey) {
+            fields.add(ROUTING_KEY + routingKey);
+            return this;
+        }
+
+        TraceBuilder queueName(String name) {
+            fields.add(QUEUE_NAME + name);
+            return this;
+        }
+
+        TraceBuilder consumerId(int id) {
+            fields.add(CONSUMER_ID + String.valueOf(id));
+            return this;
+        }
+
+        TraceBuilder fieldList(List<TraceField> traceFields) {
+            traceFields.forEach(k -> fields.add(k.getField()));
+            return this;
+        }
+
+        TraceBuilder field(TraceField traceField) {
+            fields.add(traceField.getField());
+            return this;
+        }
+
+        TraceBuilder redeliveryCount(int redeliveryCount) {
+            fields.add(REDELIVERY_COUNT + String.valueOf(redeliveryCount));
+            return this;
+        }
+
+        TraceBuilder isRedelivered(boolean redelivered) {
+            fields.add(REDELIVERED + String.valueOf(redelivered));
+            return this;
+        }
+
+        String buildTrace(String description) {
+            String fieldString = fields.stream().collect(Collectors.joining(FIELD_DELIMITER));
+            return FIELD_PREFIX + fieldString + FIELD_SUFFIX + description;
```

Comment:
Should we use a StringBuilder generally in this class for String concatenation?

Reply:
Used String Builder.